### ISSUE TITLE
removeData() when not using jquery

### ISSUE THREAD

**Issue by ceymard** [julianshapiro/velocity#629]
Hello,

`removeData()` expects keys when removing data. I don't know them as I don't know what velocity caches about my elements.

Shouldn't all the cache be destroyed when keys is not given ?

This is typically for the case when one is not using jquery at all and wants to properly unload everything velocity related.

**Comment by kenwheeler** [comment]
Can you please create a JSFiddle recreating the issue using our template, which is available here:

http://jsfiddle.net/65xCP/4/

Once I can view the reduced test case, I can more effectively isolate and address your issue.

**Comment by ceymard** [comment]
On http://jsfiddle.net/mw1h30v0/1/ I modified the template to apply a simple animation on the first div.

Once the animation is done, I call removeData() on that div without any argument but the div. However, the expando is still present (see the console.dir on the first div and on the second, which shows that the first still has jQuery4398573945 (ymmv) properties, which means the cache is still present)

This leads in my application to memory leaks since the cache is not cleared. In my case, I do not want to know what are the keys ; when I call removeData(), I expect all the cache to be cleared about this element.

This is velocity's code around line 171 ;

```javascript

 /* Custom */
    $.removeData = function (node, keys) {
        var id = node[$.expando],
            store = id && cache[id];

        if (store) {
            $.each(keys, function(_, key) {
                delete store[key];
            });
        }
    };
```

I replaced it locally to achieve full cache removal by changing to ;

```javascript
    /* Custom */
    $.removeData = function (node, keys) {
        var id = node[$.expando],
            store = id && cache[id];

        if (store) {
            if (!keys) // Here !
                delete cache[id];
            else {
                $.each(keys, function(_, key) {
                    delete store[key];
                });

            }
        }
    };
```

Should have I done differently ?

**Comment by mattwagl** [comment]
Any updates on this? I'm having a similar problem but I'm not quite sure if I am missing something in the docs. 

I'm using velocity without jQuery (inside a React-based application) and I would like to `removeData` properly after animations have been completed. However the [shimmed `removeData` function](https://github.com/julianshapiro/velocity/blob/f28c610da721048abee0be4f88a5d32d71ae8ea4/velocity.js#L172-L181) seems to be a private function. At least it does not seem to be available through `window.Velocity.Utilities and therefore it's not possible to clear the data cache from outside. So I think it boils down to these questions:

How do you `removeData` of an animated node when you're using velocity without jQuery? How can we use velocity in applications with a lot of DOM removal without causing memory leaks? The FAQ provides an answer for the jQuery case but there's no clue how to do this without jQuery. 

I believe velocity can be a great companion when you need to add JavaScript animations to complex applications where changing screens can cause a lot of DOM removal. However this internal cache seems to be a problem here. So if there's anything I can do to resolve this issue I would love to help.

Thanks for any help & tips.

**Comment by ceymard** [comment]
Use velocity.Utilities.removeData(/* your dom node */), as Utilities seems to be the jQuery shim.

I also patched internally removeData with what I pasted up there to be able to call it without argument and nuke everything cached instead of having to provide keys that I don't know.

**Comment by mattwagl** [comment]
Thanks for the hint! I applied your patch to my local Velocity version, added the `removeData` call (without the keys) to my `complete`handler and watched the memory consumption with the help of Chrome Timeline tools. 

I can confirm that after adding this patch the garbage collection can actually do its work and remove references to nodes and free up memory.

@kenwheeler Is there a chance this patch might get added to Velocity? Seems like the original[ jQuery removeData](https://api.jquery.com/jQuery.removeData/) behaves just like that. Or should I just do a fork to solve my special requirement?

**Comment by ceymard** [comment]
Maybe a pull request could be arranged :)
Le mar. 16 févr. 2016 à 12:32, Matthias Wagler <notifications@github.com> a
écrit :

**Comment by AlexGalays** [comment]
I suspect many, many people actually have GC issues because they just don't know they have to call removeData themselves; it's so counter intuitive.

**Comment by dbkr** [comment]
@ceymard: given this project is looking increasingly unmaintained, I've taken your fix and committed it to a fork at https://github.com/vector-im/velocity. We wanted this fix hosted somewhere for our project, but folks are welcome to use it. I can create a pull request if requested, but haven't since the authors have said they are unlikely to accept them.

**Comment by Rycochet** [comment]
@dbkr Actually if you'd read any of the recent issues and comments you'd see that it is being maintained - the code is needing rewritten so that we can make use of pull requests and have better testing - it is purely a case of the two of us now responsible for code getting enough time to get things to the stage that everyone can help.

**Comment by dbkr** [comment]
@Rycochet That's great news. I hope you don't mind us maintaining this fork in the meantime: we and other need this patch to fix memory leaks in our applications.

Would you reconsider uploading the source of your refactor to a branch on Github and working on that branch? It doesn't matter if it's incomplete, untidy or even nonfunctional, this would let people see the work that's happening on the refactor and allow the community to be involved. Even if it's not practical for the community to contribute at this stage, being able to see the work in progress can help us gain a better understanding of the code.

**Comment by Rycochet** [comment]
@dbkr That's the beauty of GitHub :-P

We need to get the source split up and build system working first - then it'll likely get moved onto a public repo - sadly I had plenty of time until the day I got access, and then suddenly everything started getting busy everywhere else too - hopefully we'll be able to get something visible in the next two weeks or so - all current pending changes are to do with code readability and maintenance, then will start grabbing fixes from pull requests and obvious issues - we've not got a solid timeline on the next major release (a couple of months though), but next time I get to speak to @kenwheeler I've got some ideas for internal changes that (I hope) will be well worth the version bump ;-)

See #671 for info and stuff :-)

**Comment by Rycochet** [comment]
Fixed in v1.3.0
